### PR TITLE
Initial cmake build file and fixes for Windows

### PR DESCRIPTION
**Pull request by elms**
Add cmake support and Windows improvements
* Examples working on Windows and SWTPM
  * Check command availability (including for windows blocking it).
  * update unseal to work without persistent NV storage
* fix a few spelling mistakes

### CODE REVIEW

**Comment by elms** [comment]
@haydenroche5 and @dgarske This is still a work in progress, but only expect minor fixes and clean as I test a little more on Windows. Any early feedback would be appreciated.

File: examples/native/native_test.c [hunk 1]
```diff
@@ -390,6 +393,8 @@ int TPM2_Native_TestArgs(void* userCtx, int argc, char *argv[])
         }
     }
 
+    #ifndef WOLFTPM_WINAPI
```
Comment (dgarske) [inline examples/native/native_test.c:396]: Indent is off here.

File: examples/native/native_test.c [hunk 2]
```diff
@@ -1450,13 +1474,15 @@ int TPM2_Native_TestArgs(void* userCtx, int argc, char *argv[])
         TPM2_FlushContext(&cmdIn.flushCtx);
     }
 
+#if 1 //ndef WOLFTPM_WINAPI
```
Comment (dgarske) [inline examples/native/native_test.c:1477]: Consider a cleanup here... If it should always be on.

**Review by dgarske — CHANGES_REQUESTED** [review]
Wow, great work! I like the wolfSSL submodule. The `WOLFTPM_INTERFACE` arg is awesome. Tested building on my Mac and used the IBM software TPM simulator. All works great!

The compiled applications were put into the build dir, but did you consider putting the existing locations in examples directories?

File: CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+if(WIN32)
+    # TODO: fix benchmark to compile with WIN API TBS
+    target_compile_definitions(wolftpm PRIVATE
+        "_WINDLL"
+        PUBLIC
+        "NO_TPM_BENCH"
+        )
+endif(WIN32)
+
+if("${WOLFTPM_INTERFACE}" STREQUAL "SWTPM")
```
Comment (haydenroche5) [inline CMakeLists.txt:91]: Should there be an error case if `WOLFTPM_INTERFACE` doesn't evaluate to a supported value?

File: CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+elseif("${WOLFTPM_INTERFACE}" STREQUAL "WINAPI")
+    target_compile_definitions(wolftpm PUBLIC
+        "WOLFTPM_WINAPI"
+        )
+    target_link_libraries(wolftpm tbs)
+endif("${WOLFTPM_INTERFACE}" STREQUAL "SWTPM")
+
+target_include_directories(wolftpm
+    PUBLIC
+    $<INSTALL_INTERFACE:wolfssl>
```
Comment (haydenroche5) [inline CMakeLists.txt:108]: Should the wolfssl here be wolftpm?

**Review by haydenroche5 — CHANGES_REQUESTED** [review]

File: examples/native/native_test.c (hunk 1 above)
Comment (elms) [reply examples/native/native_test.c]: Done

File: examples/native/native_test.c (hunk 2 above)
Comment (elms) [reply examples/native/native_test.c]: Done

File: CMakeLists.txt (hunk 3 above)
Comment (elms) [reply CMakeLists.txt:88]: Added error

File: CMakeLists.txt (hunk 4 above)
Comment (elms) [reply CMakeLists.txt]: yep, good catch. fixed

**Comment by elms** [comment]
> The compiled applications were put into the build dir, but did you consider putting the existing locations in examples directories?

I know we do this for wolfssl, but for cmake and windows in general I think it's considered poor practice.

**Comment by haydenroche5** [comment]
> > The compiled applications were put into the build dir, but did you consider putting the existing locations in examples directories?
> 
> I know we do this for wolfssl, but for cmake and windows in general I think it's considered poor practice.

Agree with Jeff. I did it that way in wolfSSL to make it mirror autotools as much as possible, but generally speaking about CMake, all build artifacts should stay in the build directory.

**Review by haydenroche5 — APPROVED** [review]

**Review by dgarske — CHANGES_REQUESTED** [review]
I have mixed feeling about the sub-module route. Not all users need wolfCrypt for wolfTPM. Perhaps it would be easier to assume the folders are side by side `../wolfssl` like most of us have them and/or allow the path to be provided as an environment variable?

File: README.md [hunk 5]
```diff
@@ -311,6 +311,23 @@ See `docs/SWTPM.md`
...
+if CMake support is installed. The commands below can be run in
+`Developer Command Prompt`.
+
+```
+mkdir build
+cd build
+# to use installed wolfSSL location (library and headers)
+cmake .. -DWITH_WOLFSSL=/prefix/to/wolfssl/install/
+# OR to use a wolfSSL source tree
+cmake .. -DWITH_WOLFSSL_TREE=/path/to/wolfssl/
```
Comment (dgarske) [inline README.md:326]: Using `/path/to/wolfssl/` work for me... was missing `wolfssl/options.h`. However I was able to use `/usr/local`.

File: README.md [hunk 6]
```diff
@@ -311,6 +311,23 @@ See `docs/SWTPM.md`
...
 See `docs/WindowTBS.md`
 
+## Building using CMake
+
+CMake supports compiling in many environments including Visual Studio
+if CMake support is installed. The commands below can be run in
+`Developer Command Prompt`.
+
+```
+mkdir build
```
Comment (dgarske) [inline README.md:321]: Can we add `build` to the `.gitignore`?

File: examples/seal/unseal.c [hunk 7]
```diff
@@ -85,12 +94,27 @@ int TPM2_Unseal_Example(void* userCtx, int argc, char *argv[])
     }
     printf("wolfTPM2_Init: success\n");
 
+    rc = getPrimaryStoragekey(&dev, &storage, TPM_ALG_RSA);
```
Comment (dgarske) [inline examples/seal/unseal.c:97]:
```
examples/seal/unseal.c:97:10: error: implicit declaration of function 'getPrimaryStoragekey' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
    rc = getPrimaryStoragekey(&dev, &storage, TPM_ALG_RSA);
         ^
examples/seal/unseal.c:100:10: error: implicit declaration of function 'readKeyBlob' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
    rc = readKeyBlob(inkeyfilename, &newKey);
         ^
```

**Review by dgarske — CHANGES_REQUESTED** [review]
I suspect there is some inconsistency about the WOLFTPM_AUTODETECT or conflict with the generated wolftpm/options.h... although I did try clearing it.

```
Consolidate compiler generated dependencies of target wolftpm
[  1%] Building C object CMakeFiles/wolftpm.dir/src/tpm2.c.o
/Users/davidgarske/GitHub/wolfTPM/src/tpm2.c:5252:24: error: unknown type name 'GetRandom2_In'; did you mean 'GetRandom_In'?
TPM_RC TPM2_GetRandom2(GetRandom2_In* in, GetRandom2_Out* out)
                       ^~~~~~~~~~~~~
                       GetRandom_In
/usr/local/include/wolftpm/tpm2.h:1777:3: note: 'GetRandom_In' declared here
} GetRandom_In;
  ^
/Users/davidgarske/GitHub/wolfTPM/src/tpm2.c:5252:43: error: unknown type name 'GetRandom2_Out'; did you mean 'GetRandom_Out'?
TPM_RC TPM2_GetRandom2(GetRandom2_In* in, GetRandom2_Out* out)
                                          ^~~~~~~~~~~~~~
                                          GetRandom_Out
/usr/local/include/wolftpm/tpm2.h:1780:3: note: 'GetRandom_Out' declared here
} GetRandom_Out;
  ^
/Users/davidgarske/GitHub/wolfTPM/src/tpm2.c:5265:59: error: use of undeclared identifier 'TPM_CC_GetRandom2'; did you mean 'TPM_CC_GetRandom'?
        TPM2_Packet_Finalize(&packet, TPM_ST_NO_SESSIONS, TPM_CC_GetRandom2);
                                                          ^~~~~~~~~~~~~~~~~
                                                          TPM_CC_GetRandom
/usr/local/include/wolftpm/tpm2.h:223:5: note: 'TPM_CC_GetRandom' declared here
    TPM_CC_GetRandom                = 0x0000017B,
    ^
3 errors generated.
```

File: README.md (hunk 6 above)
Comment (elms) [reply README.md:321]: Done

File: examples/seal/unseal.c (hunk 7 above)
Comment (elms) [reply examples/seal/unseal.c:98]: Fixed

File: README.md (hunk 5 above)
Comment (elms) [reply README.md:326]: Not able to reproduce any error building. Maybe older wolfTPM headers installed in the same path as wolfSSL.

File: examples/seal/unseal.c [hunk 8]
```diff
@@ -28,6 +28,7 @@
 #include <examples/seal/seal.h>
 #include <examples/tpm_io.h>
 #include <examples/tpm_test.h>
+#include <examples/tpm_test_keys.h>
```
Comment (dgarske) [inline examples/seal/unseal.c:31]: Getting a linker error. You'll need to add `tpm_test_keys.c` to the examples/seal/include.am:

```
wolfTPM % make
/Applications/Xcode.app/Contents/Developer/usr/bin/make -j17  all-am
  CCLD     examples/seal/unseal
Undefined symbols for architecture x86_64:
  "_getPrimaryStoragekey", referenced from:
      _TPM2_Unseal_Example in unseal.o
  "_readKeyBlob", referenced from:
      _TPM2_Unseal_Example in unseal.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

File: README.md (hunk 5 above)
Comment (dgarske) [reply README.md:326]: Ah, I was not using `WITH_WOLFSSL_TREE`. However then I tried it and got:

```
CMake Warning:
  Manually-specified variables were not used by the project:

    WITH_WOLFSSL_TREE
```

Then deleted the build dir and started over and it worked fine. Saw a bunch of configure like messages `-- Looking for arpa/inet.h` and `-- Looking for arpa/inet.h - found` and all built great.

**Review by dgarske — CHANGES_REQUESTED** [review]
Also having some issues with seal / unseal when built with CMake:

CMake with WOLFTPM_INTERFACE=SWTPM

```
davidgarske@Davids-MBP bin % ./seal
TPM2.0 Simple Seal example
	Key Blob: keyblob.bin
	Use Parameter Encryption: NULL
Loading SRK: Storage 0x81000200 (282 bytes)
Sealing the user secret into a new TPM key
Created new TPM seal key (pub 46, priv 141 bytes)
0x0
Wrote 193 bytes to keyblob.bin
davidgarske@Davids-MBP bin % ./unseal
Example how to unseal data using TPM2.0
wolfTPM2_Init: success
Loading SRK: Storage 0x81000200 (282 bytes)
Reading 193 bytes from keyblob.bin
Reading the private part of the key
wolfTPM2_LoadKey failed
```

configure --enable-swtpm

```
davidgarske@Davids-MBP wolfTPM % ./examples/seal/seal
TPM2.0 Simple Seal example
	Key Blob: keyblob.bin
	Use Parameter Encryption: NULL
Loading SRK: Storage 0x81000200 (282 bytes)
Sealing the user secret into a new TPM key
Created new TPM seal key (pub 46, priv 141 bytes)
0x0
Wrote 193 bytes to keyblob.bin
davidgarske@Davids-MBP wolfTPM % ./examples/seal/unseal
Example how to unseal data using TPM2.0
wolfTPM2_Init: success
Loading SRK: Storage 0x81000200 (282 bytes)
Reading 193 bytes from keyblob.bin
Reading the private part of the key
Loaded key to 0x80000002
Unsealing succeeded
Stored unsealed data to file = unseal.bin
```

File: examples/seal/unseal.c (hunk 8 above)
Comment (elms) [reply examples/seal/unseal.c:31]: Fixed

**Comment by elms** [comment]
I was failing to unload the handles. The error in unseal was when no additional handles were available. Thanks for catching that.

**Review by dgarske — APPROVED** [review]

**Pull request merged**